Sources/MapboxCoreNavigation/CoreNavigationNavigator.swift

```diff
@@ -117,13 +157,56 @@ class Navigator {
     
     var electronicHorizonOptions: ElectronicHorizonOptions? {
         didSet {
-            let nativeOptions: MapboxNavigationNative.ElectronicHorizonOptions?
-            if let electronicHorizonOptions = electronicHorizonOptions {
-                nativeOptions = MapboxNavigationNative.ElectronicHorizonOptions(electronicHorizonOptions)
-            } else {
-                nativeOptions = nil
+            setupElectronicHorizonOptions()
+        }
+    }
+}
+
+extension Navigator: FallbackVersionsObserver {
+    
+    enum TileVersionState {
+        /// No tiles version switch is required. Navigator has enough tiles for map matching.
+        case nominal
+        /// Navigator does not have tiles on current version for map matching, but TileStore contains regions with required tiles of a different version
+        case shouldFallback([String])
+        /// Navigator is in a fallback mode but newer tiles version were successefully downloaded and ready to use.
+        case shouldReturnToLatest
+    }
+    
+    func onFallbackVersionsFound(forVersions versions: [String]) {
+        switch tileVersionState {
+        case .nominal, .shouldReturnToLatest:
+            tileVersionState = .shouldFallback(versions)
+            DispatchQueue.main.async {
+                guard let fallbackVersion = versions.last else { return }
+                
+                Navigator.shared.restartNavigator(forcing: fallbackVersion)
+                
+                let userInfo: [Navigator.NotificationUserInfoKey: Any] = [
+                    .tilesVersionKey: fallbackVersion
+                ]
+                
+                NotificationCenter.default.post(name: .navigationDidSwitchToFallbackVersion,
+                                                object: nil,
+                                                userInfo: userInfo)
             }
-            navigator.setElectronicHorizonOptionsFor(nativeOptions)
+        case .shouldFallback:
+            break // do nothing
+        }
+    }
+
+    func onCanReturnToLatest() {
+        switch tileVersionState {
+        case .nominal, .shouldFallback:
+            tileVersionState = .shouldReturnToLatest
```

Comment:
Probably makes sense to read/mutate `tileVersionState` on the main thread as well. It is very easy to introduce a data race when variables mutates and read from different threads.